### ISSUE TITLE
Time clock addon?

### ISSUE THREAD

**Issue by odiea** [opensourcepos/opensourcepos#2825]
Just a quick question. If someone is familiar with time math would it be possible to create a time clock module using the current Cashups module? It has a open and close and would need a reason for logging in and out say lunch,Break Clock . Just an idea.

**Comment by odiea** [comment]
Or would expenses be better. It already has the expenses categories that could be converted to reasons for clocking in and out.

**Comment by WebShells** [comment]
@odiea i think great ideas to start with if all what we are after is manual clock in method with a simple report calculating different and summing logged in hours.

**Comment by odiea** [comment]
How about using Expenses with an app config similar to dinner tables or is app config getting too overloaded now?

**Comment by WebShells** [comment]
@odiea App config is already divided into 2 sections by the last PR i submitted so we can add TABs without any issues or we can merge what can be merged together.

**Comment by WebShells** [comment]
Please check the demo

**Comment by odiea** [comment]
I did and it looks good. I am still deciding what to do. Expenses Categories looks to me like the one I will try first.

**Comment by odiea** [comment]
I stuck with using the expenses and expense categories as a template. All is working good now. I am seeing no issues or errors. The hard part is going to be writing a report using the following db. Now should DateTime be a single row as it is now or should a time row be added?
![2020-05-09 12_21_41-Window](https://user-images.githubusercontent.com/10962177/81481863-455b2080-91f0-11ea-8aab-2549bd94b997.png)

**Comment by WebShells** [comment]
@odiea If you are planning/interested to develop this further, i would be glad to write down the scenario and whats the best for it to be working as expected including reports, etc... I guess for your question, no need to be separated into 2 columns; the timestamp needed is Date + Timing

**Comment by odiea** [comment]
I have gotten as far as Creating and adding Times and creating and adding Time categories. It has the look and fell of expenses:) Now your suggestions would be appreciated.

**Comment by WebShells** [comment]
@odiea I can see the basics/logic now are there.. We might need to consider the below in order to get the best out of it:

1) Clock in / Clock out Categories, ex:
- Clock-in a shift before starting ( Category: AM Shift, PM Shift )
- Clock-out Categories ( Sick Leave, Break, Yearly Vacation, Etc.. )

2)  Automatic Clock-in on Login
Maybe a modal should popup if the employee isn't currently clocked in yet, with the list of clock-in Categories. 
Same to happen on logout, with a list of clock OUT categories as dropdown.
3) Clock in-outs should state IP addresses of the employees not to logout from outside the business location.
4) Hourly Rates
If we have the chance to add a field to Employees tables ( Hourly Rate ) upon adding an employee. This will be used for reports calculation, in order to calculate the salaries as hours * hourly rate and display it in the Attendance Detailed Report. 
These are the main things that every business needs to fully take advantage of this module.

For the reporting section:

Graphical:
Will show a graph including employee names, total number of hrs and % of work load on each

Summary:
Employee | Hours | Total $

Detailed:
+ Employee | Clock-in | Clock-out | Hours | Total 
-  Clock in Category | Clock out Category | Hourly Rate | Clock-in IP Address | Clock-Out IP Address
The above detailed report to be capable of editing by Manager or deleting any in-valid clock-in, etc...

Finally option to print tables as already set for other sections. 
Maybe am missing something, we will reach up the right results when we get on it :)...
Thank you

**Comment by odiea** [comment]
1. The ability to add Categories is available in the Time Clock Categories and Time clock has a dropdown to add any reason that was added in The Categories.
2. Yes I think that would or should be added after login.
3. It looks like an ip address and hourly rate could be added to the current db?
Reports would be great.

**Comment by odiea** [comment]
If you would like I could zip up what I have so far for you to check out?

**Comment by WebShells** [comment]
@odiea sure, will be glad to

**Comment by odiea** [comment]
I believe that this contains all the new info. If you want to use it just add it to addons. I forgot to do that on my first try.
[timeclock.zip](https://github.com/opensourcepos/opensourcepos/files/4605856/timeclock.zip)

**Comment by odiea** [comment]
I forgot the lang files. Let me know if you would like them. Sorry En-Us only.

**Comment by WebShells** [comment]
It's fine

**Comment by WebShells** [comment]
Views are missing @odiea ?

**Comment by odiea** [comment]
I knew I was forgetting something.
[timeclock.zip](https://github.com/opensourcepos/opensourcepos/files/4605960/timeclock.zip)

**Comment by WebShells** [comment]
ERROR - 2020-05-10 18:49:23 --> Severity: error --> Exception: Call to undefined function get_timeclocks_manage_table_headers() 
In controllers...

**Comment by odiea** [comment]
That is in the file timeclock tabular helper and needs to be merged with Tabular_helper

**Comment by WebShells** [comment]
Same error, unless am doing something wrong...

**Comment by odiea** [comment]
Here is my complete tabular_helper. See if this helps.
[tabular_helper.zip](https://github.com/opensourcepos/opensourcepos/files/4606162/tabular_helper.zip)

**Comment by WebShells** [comment]
ERROR - 2020-05-10 20:34:47 --> Severity: error --> Exception: Call to undefined function ch_title() \views\timeclocks\manage.php 70
ERROR - 2020-05-10 20:34:48 --> Severity: error --> Exception: Call to a member function search() on null controllers\Timeclocks.php 32
ERROR - 2020-05-10 20:34:49 --> Severity: error --> Exception: Call to a member function get_info() on null controllers\Timeclocks.php 60

**Comment by odiea** [comment]
I guess if you are brave you can see it on my test computer. Not https so you will get a warning. www.ospos.25u.com

**Comment by odiea** [comment]
Sorry I was not showing any errors either in the log or 2 browsers using inspect.

**Comment by WebShells** [comment]
@odiea the date - time shouldn't be selected by the employee also it should be automatically logging the clock in timestamp. I wonder why it didn't work on my local install. I still prefer our current theme more than AdminLTE ^_^

**Comment by tmn14** [comment]
Hi @odiea i agree with @WebShells , the date and time shouldn't be selected by the employee.
@WebShells regarding your basics/logic this is how i see it:
1. clock in Categories its a good idea (i think it will be best if the categories will be in the configuration file instead of  Time Clock Categories like expense) and we can redirect base on login/logout that the employee select the correct option.
2. if the clock is base on login and we redirect him to choose an option before loading the home models it will never be empty, we need to consider upon missing logout (user my close the browser and not logout properly or for timeout session we need to check upon login if the last shift not clocked out..
3. good thought regard the ip can be done upon login.
4. this will be great If we can like you suggested  add a field to Employees tables ( Hourly Rate )

this is can be a great addon for ospos !

**Comment by odiea** [comment]
@WebShells and @tmn14 Thanks for the input. I already changed the calendar to read only so 1 problem is resolved. I will add the hr rate and ip address to the timeclocks db and go from there. For me I will keep the categories in the timeclock categories and place it into the office module so a Clerk or Cashier has not access to it.

**Comment by tmn14** [comment]
@odiea >

>  place it into the office module

Sounds like a good solution if you need a hand I'd be happy to help

**Comment by odiea** [comment]
@WebShells . Does your application/config/autoload contain the following under autoload model?
'Timeclock',
'Timeclock_category'

**Comment by WebShells** [comment]
@odiea i added them yes, let me try to use latest master am using an older one

**Comment by odiea** [comment]
No problem. Just trying to figure out why it is not working for you. The following error is due to the footer being altered so you can comment that out or remove it.
ERROR - 2020-05-10 20:34:47 --> Severity: error --> Exception: Call to undefined function ch_title() \views\timeclocks\manage.php 70

**Comment by WebShells** [comment]
let's clean up the issue to keep the conv of main topic, we messed it up ;p

**Comment by odiea** [comment]
@WebShells  Please contact me for further discussion on this issue

**Comment by WebShells** [comment]
@odiea thanks for the editions you've made here... We'd like to see @opensourcepos/org comment on this to know if we should proceed or use it without adding/modifying it and merging to master...

**Comment by WebShells** [comment]
If yes, we apply the comments above and submit a PR

**Comment by odiea** [comment]
Did you get it to work for you?

**Comment by WebShells** [comment]
Yes i didnt have the time to play around with the code, but i got it work on the latest master

**Comment by odiea** [comment]
I did make 2 changes. I added the ip address and made the employee dropdown disabled.

**Comment by WebShells** [comment]
@daN4cat @jekkos @SteveIreland can you add your comments regarding this please ? Would like to know your opinion of what's being cooked

**Comment by WebShells** [comment]
@odiea was checking your test version now upon login it takes to timeclock which is how it should work in case the user wasn't logged in... Am asking how pro we are going to make this module because if we want to do it as it should be, we should be disabling clock in option if the user is Clocked IN and only display Clock Out option or break or whatever we define. Also we should display the IP address with IN/OUT time in the same table ( in time clocks )... In the tabular view the columns should be displaying as follows: 
ID | Clock IN Time | Clock Out Time | Desc | Employee | Edit
Let's all agree on the scenario so we can divide the work

**Comment by odiea** [comment]
Sounds good. Play with your version and let me know how it works out. Thanks

**Comment by odiea** [comment]
I knew that this so far is just the rough basic draft so nothing is set in stone.

**Comment by odiea** [comment]
The biggest issue is going to be multiple logins and outs. Do we do anything with the lunch break or coffee break?

**Comment by odiea** [comment]
I am using the following code for logout. Is there a way to redirect it to the timeclock if cancelled is pressed?
```
<a href="<?php echo site_url("home/logout");?>" onclick="return confirm('Did you clock out? If not Cancel and Clock Out');"><i class="fas fa-sign-out-alt"></i><?php echo $this->lang->line("common_logout")?></a>									
```
